File: apps/mobile-passenger/src/lib/geo.ts
import type { Coords } from '../services/location';

const EARTH_RADIUS_KM = 6371;

/** Distancia em km (Haversine). */
export function distanceKm(a: Coords, b: Coords): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(b.latitude - a.latitude);
  const dLon = toRad(b.longitude - a.longitude);
  const lat1 = toRad(a.latitude);
  const lat2 = toRad(b.latitude);
  const h = Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(h));
}

/** Interpola um ponto entre a e b (t de 0 a 1) com leve curva. */
export function interpolate(a: Coords, b: Coords, t: number, curve = 0): Coords {
  const mid = {
    latitude: (a.latitude + b.latitude) / 2 + curve * (b.longitude - a.longitude) * 0.12,
    longitude: (a.longitude + b.longitude) / 2 - curve * (b.latitude - a.latitude) * 0.12,
  };
  const ab = (1 - t) ** 2;
  const mb = 2 * (1 - t) * t;
  const bb = t ** 2;
  return {
    latitude: ab * a.latitude + mb * mid.latitude + bb * b.latitude,
    longitude: ab * a.longitude + mb * mid.longitude + bb * b.longitude,
  };
}

/** Gera pontos de uma rota simulada entre origem e destino. */
export function buildRoute(from: Coords, to: Coords, steps = 40): Coords[] {
  return Array.from({ length: steps + 1 }, (_, i) => interpolate(from, to, i / steps, 1));
}

/** Converte metros para texto legivel. */
export function formatDistance(meters: number): string {
  return meters < 1000 ? `${Math.round(meters)} m` : `${(meters / 1000).toFixed(1)} km`;
}

/** Converte segundos para "X min". */
export function formatDuration(seconds: number): string {
  const minutes = Math.max(1, Math.round(seconds / 60));
  return `${minutes} min`;
}

/** Converte centavos para "R$ 0,00". */
export function formatMoney(cents: number): string {
  return (cents / 100).toLocaleString('pt-BR', { style: 'currency', currency: 'BRL' });
}

export function formatDateTime(iso: string): string {
  const date = new Date(iso);
  return date.toLocaleString('pt-BR', { day: '2-digit', month: '2-digit', hour: '2-digit', minute: '2-digit' });
}

File: apps/mobile_driver/lib/core/utils/geo.dart
import 'dart:math' as math;

/// Coordenada geografica simples (evita acoplar o dominio ao plugin de GPS).
class Coords {
  const Coords(this.latitude, this.longitude);

  final double latitude;
  final double longitude;

  @override
  String toString() => '${latitude.toStringAsFixed(5)}, ${longitude.toStringAsFixed(5)}';

  Map<String, double> toJson() => {'latitude': latitude, 'longitude': longitude};
}

/// Centro inicial do mapa (Sao Paulo), usado enquanto o GPS nao responde.
// Valor TECNICO inicial, nunca mostrado: os mapas so abrem com a posicao
// real do aparelho (tela "Localizando voce"). Nada de cidade fixa na tela.
const Coords fallbackCoords = Coords(-18.0125, -49.3547);

const double _earthRadiusKm = 6371;

double _toRad(double deg) => deg * math.pi / 180;

/// Distancia em km entre dois pontos (Haversine).
double distanceKm(Coords a, Coords b) {
  final dLat = _toRad(b.latitude - a.latitude);
  final dLon = _toRad(b.longitude - a.longitude);
  final lat1 = _toRad(a.latitude);
  final lat2 = _toRad(b.latitude);
  final h = math.pow(math.sin(dLat / 2), 2) +
      math.cos(lat1) * math.cos(lat2) * math.pow(math.sin(dLon / 2), 2);
  return 2 * _earthRadiusKm * math.asin(math.sqrt(h.toDouble()));
}

/// Ponto intermediario com leve curva (para uma rota visualmente natural).
Coords interpolate(Coords a, Coords b, double t) {
  final mid = Coords(
    (a.latitude + b.latitude) / 2 + (b.longitude - a.longitude) * 0.12,
    (a.longitude + b.longitude) / 2 - (b.latitude - a.latitude) * 0.12,
  );
  final ab = math.pow(1 - t, 2).toDouble();
  final mb = 2 * (1 - t) * t;
  final bb = math.pow(t, 2).toDouble();
  return Coords(
    ab * a.latitude + mb * mid.latitude + bb * b.latitude,
    ab * a.longitude + mb * mid.longitude + bb * b.longitude,
  );
}

/// Gera os pontos de uma rota simulada entre origem e destino.
List<Coords> buildRoute(Coords from, Coords to, {int steps = 40}) {
  return List<Coords>.generate(steps + 1, (i) => interpolate(from, to, i / steps));
}

/// "820 m" ou "4,3 km".
String formatDistance(double meters) {
  if (meters < 1000) return '${meters.round()} m';
  final km = meters / 1000;
  return '${km.toStringAsFixed(1).replaceAll('.', ',')} km';
}

/// "12 min".
String formatDuration(int seconds) {
  final minutes = math.max(1, (seconds / 60).round());
  return '$minutes min';
}

/// Normaliza a placa: maiuscula, sem separadores.
String normalizePlate(String value) =>
    value.toUpperCase().replaceAll(RegExp(r'[^A-Z0-9]'), '');

/// Valida a placa no padrao antigo (ABC1234) ou Mercosul (ABC1D23).
bool isValidPlate(String value) {
  final plate = normalizePlate(value);
  return RegExp(r'^[A-Z]{3}[0-9]{4}$').hasMatch(plate) ||
      RegExp(r'^[A-Z]{3}[0-9][A-Z][0-9]{2}$').hasMatch(plate);
}

/// Formata a placa para exibicao: ABC-1234 ou ABC1D23.
String formatPlate(String value) {
  final plate = normalizePlate(value);
  if (RegExp(r'^[A-Z]{3}[0-9]{4}$').hasMatch(plate)) {
    return '${plate.substring(0, 3)}-${plate.substring(3)}';
  }
  return plate;
}
